### PR TITLE
Improve SnapshotInfo serialization tests

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we have no tests that directly verify that `SnapshotInfo`
correctly round-trips through its in-repository `XContent`-based
representation. We do have tests that verify that it correctly
round-trips through the wire format, but these tests always use an empty
collection for `featureStates`.

This commit strengthens the wire format round-trip tests to use
nontrivial `featureStates`, and generalizes these tests to also verify
that the `XContent` representation is faithful.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+public class SnapshotInfoTestUtils {
+    private SnapshotInfoTestUtils() {
+    }
+
+    static SnapshotInfo createRandomSnapshotInfo() {
+        final SnapshotId snapshotId = new SnapshotId(randomAlphaOfLength(5), randomAlphaOfLength(5));
+        final List<String> indices = Arrays.asList(randomArray(1, 10, String[]::new, () -> randomAlphaOfLengthBetween(2, 20)));
+
+        final List<String> dataStreams = Arrays.asList(randomArray(1, 10, String[]::new, () -> randomAlphaOfLengthBetween(2, 20)));
```
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoTestUtils.java:41]: Maybe inject random empty data streams? (unless it complicates too much the random mutations)

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoXContentTests.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.PlainActionFuture;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
+import org.elasticsearch.test.AbstractWireTestCase;
+
+import java.io.IOException;
+
+public class SnapshotInfoXContentTests extends AbstractWireTestCase<SnapshotInfo> {
```
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoXContentTests.java:21]: It seems that SnapshotInfo is lenient when it comes to parse unknown fields. Should we add some AbstractXContentTestCase just for SnapshotInfo to test this too?

**Review by tlrx — APPROVED** [review]
LGTM. I left minor comments, feel free to ignore if needed.

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoTestUtils.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoTestUtils.java:41]: Good point, that's easy enough, see f061963.

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoXContentTests.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoXContentTests.java]: Unfortunately we don't seem to parse it as `AbstractXContentTestCase` requires today. Likely fixable but I didn't really want to make production code changes here. At least we have good tests for the behaviour of repositories after upgrades and downgrades which should catch any problems in this area.

Comment (tlrx) [reply server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoXContentTests.java]: Sure, let's move on.

Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/snapshots/SnapshotInfoXContentTests.java]: It turns out we do have some coverage of this area via `GetSnapshotsResponseTests`; improvements are coming in https://github.com/elastic/elasticsearch/pull/72079.

**Pull request merged**